server/src/storage/time_series.rs

```diff
@@ -0,0 +1,212 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! Support for managing multiple compressed segments from the same time series.
+
+use std::fs;
+use std::io::ErrorKind::Other;
+use std::sync::Arc;
+
+use datafusion::arrow::datatypes::{ArrowPrimitiveType, DataType, Field, Schema};
+use datafusion::arrow::record_batch::RecordBatch;
+
+use crate::storage::write_batch_to_parquet;
+use crate::types::{ArrowTimestamp, ArrowValue, TimestampArray};
+
+/// A single compressed time series, containing one or more compressed segments and providing
+/// functionality for appending segments and saving all segments to a single Parquet file.
+pub struct CompressedTimeSeries {
+    /// Compressed segments that make up the sequential compressed data of the time series.
+    compressed_segments: Vec<RecordBatch>,
+    /// Continuously updated total sum of the size of the compressed segments.
+    pub size_in_bytes: usize,
+}
+
+impl CompressedTimeSeries {
+    pub fn new() -> Self {
+        Self {
+            compressed_segments: Vec::new(),
+            size_in_bytes: 0,
+        }
+    }
+
+    /// Append `segment` to the compressed data in the time series and return the size of `segment` in bytes.
+    pub fn append_segment(&mut self, segment: RecordBatch) -> usize {
+        let segment_size = Self::get_size_of_segment(&segment);
+
+        debug_assert!(
+            segment.schema() == Arc::new(Self::get_compressed_segment_schema()),
+            "Schema of record batch does not match compressed segment schema."
+        );
+
+        self.compressed_segments.push(segment);
+        self.size_in_bytes += segment_size;
+
+        segment_size
+    }
+
+    /// If the compressed segments are successfully saved to Parquet, return Ok, otherwise return Err.
+    pub fn save_to_parquet(&mut self, key: String) -> Result<(), std::io::Error> {
+        if self.compressed_segments.is_empty() {
+            Err(std::io::Error::new(
+                Other,
+                "The compressed time series does not contain any compressed data.",
+            ))
+        } else {
+            // Combine the segments into a single record batch.
+            let schema = Self::get_compressed_segment_schema();
+            let batch = RecordBatch::concat(&Arc::new(schema), &*self.compressed_segments).unwrap();
+
+            // Create the folder structure if it does not already exist.
+            let folder_path = format!("storage/{}/compressed", key);
+            fs::create_dir_all(&folder_path)?;
+
+            // Create a path that uses the first timestamp as the filename.
+            let start_times: &TimestampArray = batch.column(2).as_any().downcast_ref().unwrap();
```

Comment:
Is the first timestamp used with indexing in mind (e.g., pruning data that is too new or old  to be part of the query) if so it should probably be documented?

Reply:
Technically it was just used since it was unique but i have mentioned the thing about pruning in the comment.